### PR TITLE
Create spec for lambda params with modifiers without type name

### PR DESCRIPTION
**Pull request by Rekkonnect**
Creates a spec for #338

This permits the simplified declaration of lambdas with parameters with modifiers, by omitting their type name.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Can you add a line saying this is only for parenthesized lambdas?

File: proposals/ref-out-lambda-params.md
Comment (CyrusNajmabadi) [inline proposals/ref-out-lambda-params.md:26]: Note. We may have to also support `ref readonly ` as that feature is currently being worked on. If that impacts delegates, it would impact this.  @jjonescz ?

File: proposals/ref-out-lambda-params.md
Comment (CyrusNajmabadi) [inline proposals/ref-out-lambda-params.md:52]: This rule seems superfluous given the one below.

File: proposals/ref-out-lambda-params.md
Comment (CyrusNajmabadi) [inline proposals/ref-out-lambda-params.md:57]: Sentence seems superfluous. You already mentioned this above.

File: proposals/ref-out-lambda-params.md
Comment (CyrusNajmabadi) [inline proposals/ref-out-lambda-params.md:82]: Could just inline this method to simplify the example.

**Comment by CyrusNajmabadi** [comment]
Speclet lgtm.  @333fred for another pair of eyes.

File: proposals/ref-out-lambda-params.md
Comment (Rekkonnect) [reply proposals/ref-out-lambda-params.md:82]: In this example I'm trying to showcase that, even though the actual parameter types can be inferred from the usage of the parameters inside the body (through `Method2`), they are not being inferred. If that was not clear enough, I can mention the intend of the example.

Comment (CyrusNajmabadi) [reply proposals/ref-out-lambda-params.md:79]: wfm.

File: proposals/ref-out-lambda-params.md
Comment (Rekkonnect) [reply proposals/ref-out-lambda-params.md:26]: So do we also declare the support for `ref readonly` in the spec before its approval?

Comment (CyrusNajmabadi) [reply proposals/ref-out-lambda-params.md:26]: Ref-readonly is approved and shipping for 12.  Subsequent proposals should incorporate the fallout from them. :-)

**Comment by Rekkonnect** [comment]
An edge case includes the addition of the `scoped` modifier in C# 11. Currently, lambda parameters accept the scoped modifier, therefore those could also be affected from this spec. We want to decide whether those are being supported or not.

**Comment by Rekkonnect** [comment]
Also, in preview, `params` in lambda parameters is being supported. We also need to consider this case.

File: proposals/ref-out-lambda-params.md
Comment (333fred) [inline proposals/ref-out-lambda-params.md:43]:
```suggestion
The change in the spec will require that [the grammar for lambda expressions](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#12191-general) be adjusted as follows:
```

File: proposals/ref-out-lambda-params.md
Comment (333fred) [inline proposals/ref-out-lambda-params.md:54]: Nit: I'd prefer to define this in terms of the existing parameter grammar, rather than copying the content of the parameter.

```suggestion
+     : anonymous_function_parameter_modifier? implicit_anonymous_function_parameter
```

File: proposals/ref-out-lambda-params.md
Comment (333fred) [inline proposals/ref-out-lambda-params.md:60]: I'm not sure what "Similarly, `async` lambdas will also not be affected from this change." means here. They will be affected by this change; it still will be illegal to have `RefFunc<int, Task> f = async (ref i) => { ... };`, but not because the grammar forbids the syntax, it's a semantic error just like having async method take ref parameters is.

**Review by 333fred — COMMENTED** [review]
A few small comments, overall this is looking good.

**Review by 333fred — COMMENTED** [review]
@CyrusNajmabadi this looks good enough to go to LDM to me. Want to add it to "schedule when convenient"?

**Comment by CyrusNajmabadi** [comment]
@333fred I was going to check with @MadsTorgersen @KathleenDollard first when teh right time was to bring things since we're so heads down in 12 right now.  Basically, i'd bring this the next time we're open to looking for the next batch of championed constructs to take :)

**Comment by 333fred** [comment]
I believe that's what "schedule when convenient" there for 😉.

**Comment by CyrusNajmabadi** [comment]
Ah, i thought those were for meetings where we didn't have anything pressing, but we were still prior to the point that we were trying to decide on new features :)

**Comment by Rekkonnect** [comment]
> As for `scoped` parameters, since `scoped` is contextual, the compiler currently attempts to parse that as a type. Perhaps we will not make a move to support `(scoped x)` to infer the type of `x` and granting it the `scoped` modifier.

As per the [list of breaking changes](https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20DotNet%207.md#types-cannot-be-named-scoped) for C# 11, `scoped` no longer refers to a type name, therefore this restriction should not take place. `scoped` can be safely evaluated as a modifier in an implicit type lambda parameter list.

Therefore:
```csharp
private delegate int D(scoped int x, scoped int y);
D d = (scoped x, scoped y) => x + y;
```
is valid and `scoped` reflects the modifier, not a type named `scoped`.

Will update the spec accordingly to reflect this change, citing the above reference.

**Comment by jjonescz** [comment]
> `scoped` no longer refers to a type name

Is that right? Seems the breaking change only disallows declaring new types named `scoped`, but you could reference an existing type named `scope` from a DLL and then this would be a break (not saying the break wouldn't be acceptable):

```cs
var d = (scoped x, scoped y) => x + y;
```

**Comment by jjonescz** [comment]
> ```cs
> ref scoped local; // Error
> ```

Maybe that's an error only when there's `ref`.

> Hints that all available types named `scoped` are now inaccessible. 

This test succeeds for me:

```cs
[Fact]
public void Scoped()
{
    var source1 = """
        public class scoped { }
        """;

    CreateCompilation(source1).VerifyDiagnostics(
        // (1,14): error CS9062: Types and aliases cannot be named 'scoped'.
        // public class scoped { }
        Diagnostic(ErrorCode.ERR_ScopedTypeNameDisallowed, "scoped").WithLocation(1, 14));

    var comp1 = CreateCompilation(source1, parseOptions: TestOptions.Regular10).VerifyDiagnostics(
        // (1,14): warning CS8981: The type name 'scoped' only contains lower-cased ascii characters. Such names may become reserved for the language.
        // public class scoped { }
        Diagnostic(ErrorCode.WRN_LowerCaseTypeName, "scoped").WithArguments("scoped").WithLocation(1, 14));
    var comp1Ref = comp1.EmitToImageReference();

    var source2 = """
        var d = (scoped x) => System.Console.WriteLine(x);
        d(new scoped());
        """;
    CompileAndVerify(source2, new[] { comp1Ref }, expectedOutput: "scoped").VerifyDiagnostics();
}
```

**Comment by Rekkonnect** [comment]
Nice observation, with a little bit more testing, even the listed breaking change does not currently apply:
[SharpLab link](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ6DEUIBtcCGARrgKYAEAJgJYDOxZ66tAxgPYAOpl5AZuQF5yOfAG50YUv1aduffkMn9e4tDK6UAdADlStYHKiDheXKqXl13HXoM8oRxVOGr0icnDhXK6AN7pyQI8AZg8Ecl19OV9yAF90WKA===)

Note that the above uses the Default branch which is C# 11 based on the fact that the error about naming a type `scoped` exists.

This is either an undocumented decision, or an omission in the feature. We need a definite answer to how we're going to support `scoped` in implicit lambda parameter lists.

**Comment by CyrusNajmabadi** [comment]
if we're going to support `ref` and the like without an explicit type, then we need to support `scoped` in the same fashoin.

**Comment by CyrusNajmabadi** [comment]
I would put it as a case to consider and allow the ldm to make a determination on if it's ok or not.  

Personally, I would take the break.

**Comment by Rekkonnect** [comment]
Digging through the issues, I also discovered that #6149 is touching the subject of `scoped`, perhaps it would be helpful to mention it here for future reference once the spec is settled.

File: proposals/ref-out-lambda-params.md
Comment (CyrusNajmabadi) [inline proposals/ref-out-lambda-params.md:28]:
```suggestion
For example, these would not be legal.
```

Also, just do like the first 3 examples.

**Review by CyrusNajmabadi — APPROVED** [review]

File: proposals/ref-out-lambda-params.md
Comment (Rekkonnect) [inline proposals/ref-out-lambda-params.md:111]: I think this note should also be removed

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**
